lib/grp.gi

```diff
@@ -229,9 +233,13 @@ InstallMethod( IsPGroup,
         if ord = infinity then
           return false;
         elif 1 < ord then
-          UniteSet( s, Factors( ord ) );
-          if 1 < Length( s ) then
+          if not IsPrimePowerInt( ord ) then
             return false;
+          else
+            AddSet( s, Factors( ord )[1] );
+            if 1 < Length( s ) then
+              return false;
+            fi;
           fi;
```

Comment:
Can't you avoid the code duplication by moving the common code into a separate function, which you then call from two places? That is the traditional way, and it is probably easier to understand, reason about, and debug, than using `RedispatchOnCondition`....